ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -582,6 +582,80 @@ public void testRangeRequestCompositeBlob()
   }
 
   /**
+   * Test that the operation is completed and an exception with the error code {@link RouterErrorCode#ChannelClosed} is
+   * set when the {@link ReadableStreamChannel} is closed before all chunks are read.
+   * @throws Exception
+   */
+  @Test
+  public void testEarlyReadableStreamChannelClose()
+      throws Exception {
+    for (int numChunksInBlob = 0; numChunksInBlob <= 4; numChunksInBlob++) {
+      for (int numChunksToRead = 0; numChunksToRead < numChunksInBlob; numChunksToRead++) {
+        testEarlyReadableStreamChannelClose(numChunksInBlob, numChunksToRead);
+      }
+    }
+  }
+
+  /**
+   * Test that the operation is completed and an exception with the error code {@link RouterErrorCode#ChannelClosed} is
+   * set when the {@link ReadableStreamChannel} is closed before all chunks are read for a specific blob size and
+   * number of chunks to read.
+   * @param numChunksInBlob the number of chunks in the blob to put/get.
+   * @param numChunksToRead the number of chunks to read from the {@link AsyncWritableChannel} before closing the
+   *                        {@link ReadableStreamChannel}.
+   * @throws Exception
+   */
+  private void testEarlyReadableStreamChannelClose(int numChunksInBlob, final int numChunksToRead)
+      throws Exception {
+    final AtomicReference<Exception> callbackException = new AtomicReference<>();
+    final CountDownLatch readCompleteLatch = new CountDownLatch(1);
+    Callback<GetBlobResult> callback = new Callback<GetBlobResult>() {
+      @Override
+      public void onCompletion(final GetBlobResult result, Exception exception) {
+        if (exception != null) {
+          callbackException.set(exception);
+          readCompleteLatch.countDown();
+        } else {
+            final ByteBufferAsyncWritableChannel writableChannel = new ByteBufferAsyncWritableChannel();
+            result.getBlobDataChannel().readInto(writableChannel, null);
```

Comment:
Shouldn't this `readInto()` get the `ClosedChannelException` (if the future it returns is checked after the operation completes)? If so, could we check for that at the end as well?